content/blog/2019-11-01_react-typescript-webpack-2.md

```diff
@@ -0,0 +1,193 @@
+This means that every time someone changes anything in the codebase, the full bundle will be invalidated and re-downloaded by every users.
+
+There are two approaches to splitting the chunks:
+
+1. **bundle splitting**: separating into fixed bundles, for example putting dependencies into another bundle
+2. **code splitting**: splitting your own code into multiple bundles loaded on demand
+
+### Bundle splitting
+The easiest bundle splitting strategy is to create a `vendors` bundle containing all dependencies. In our case we currently
+only have `react` and `react-dom` but it will inevitably grow and, unless you are upgrading dependencies every day, will not
+change very often. Dividing the bundle does not change anything for first-time users as they will have to download all files but
+repeat users will only download the bundles that changed.
+
+Webpack comes with a built-in option to split everything coming from the `node_modules` folder to another bundle: [optimization.splitChunks.chunks](https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkschunks).
+Setting it to `"all"` will produce a new JavaScript file with a name starting by `vendors~app` in the output directory weighting 131KB while the app bundle shrank down to 1.9KB.
+If you run `build:prod` again after making a change to `index.tsx`, you will notice the hash did change: we need to ensure the hash
+is done on the content rather than the build. In practice this means changing `[hash]` to `[contenthash]` in the `output.filename` configuration in production.
+Trying it again will now give the expected results: a change in our application code doesn't change the `vendors` bundle.
+
+For many apps, just splitting the bundle that way will be enough to get started. If needed, you can create different kind
+of strategies, for example:
+
+- one file per `npm` dependency
+- group packages per update of frequency: put the ones that change often in the same bundle
+- group packages per kind: all React packages could go in one bundle, all data viz in another etc
+- have multiple entry points: Webpack will create one file per entry point
+
+The [Webpack documentation](https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkscachegroups) 
+has some examples demonstrating how to implement some of the strategies. Once again, start with the default dumb splitting and experiment as you go when needed.
+
+The commit is <https://github.com/Keats/webpack-react-typescript/commit/cd2669b62d9542794fdc52e6e6fb1702716ddc36>.
+
+### Code splitting
+Code splitting is allowing you to import some code on demand. It relies on the [dynamic import() proposal](https://github.com/tc39/proposal-dynamic-import) which
+is now on stage 4, eg finished. A simple React example would be:
+
+```tsx
+import React from "react";
+
+class LocationForm extends React.Component<{}> {
+   handleClickOnMap = () => {
+     import('./locationModal')
```

Comment:
I don't like that xD

I forgot a bit about all that. Looking at some old code, I was using React.lazy and Suspense to import files at the top of the page

```ts
const Settings = React.lazy(() => import("./pages/Settings"));

const Routes: FC<void> = () => (
  <Switch>
    <Route
      exact
      path={getSettingsUrl()}
      component={LazyComponentCreator(Settings)}
    />
  </Switch>
);

export default Routes;
```

LazyCompanyCreator:
```ts
import React, {FC, Suspense} from "react";

type Type = (Component) => FC;

/* eslint-disable-next-line react/display-name */
const LazyComponentCreator: Type = ComposedComponent => props => (
  <Suspense fallback={<div>Loading...</div>}>
    <ComposedComponent {...props} />
  </Suspense>
);

export default LazyComponentCreator;
```

Reply:
J'utilisais aussi React.memo, me semble que ça avait aussi un effet:

```ts
import React, {FC, memo} from "react";

const Top: FC = () => (
  <FlexRow center="xs" middle="xs" className="faq-top">
  </FlexRow>
);
export {Top as TopNaked};

const MemoTop = memo(Top);
/* eslint-disable-next-line immutable/no-mutation */
MemoTop.displayName = "MemoTop";
export default MemoTop;
```

Reply:
Can't say my code makes things simpler :/

Reply:
oui je dis de lire la doc de React pour code splitting et c'est en gros React.lazy. `import()` va faire/fait partie de ECMAScript par contre donc c'est là pour un bout de temps.